baseband/guppi/payload.py

```diff
@@ -0,0 +1,247 @@
+# Licensed under the GPLv3 - see LICENSE
+"""Payload for GUPPI format."""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+import numpy as np
+from collections import namedtuple
+
+from ..vlbi_base.payload import VLBIPayloadBase
+
+
+__all__ = ['GUPPIPayload']
+
+
+def decode_8bit(words):
+    return words.view(np.int8, np.ndarray).astype(np.float32)
+
+
+def encode_8bit(values):
+    return np.clip(np.rint(values), -128, 127).astype(np.int8)
+
+
+class GUPPIPayload(VLBIPayloadBase):
+    """Container for decoding and encoding GUPPI payloads.
+
+    Parameters
+    ----------
+    words : `~numpy.ndarray`
+        Array containg LSB unsigned words (with the right size) that
+        encode the payload.
+    header : `~baseband.dada.GUPPIHeader`
+        Header that provides information about how the payload is encoded.
+        If not given, the following arguments have to be passed in.
+    bps : int, optional
+        Number of bits per sample part (i.e., per channel and per real or
+        imaginary component).  Default: 8.
+    sample_shape : tuple, optional
+        Shape of the samples; e.g., (nchan,).  Default: ().
+    complex_data : bool, optional
+        Whether data are complex.  Default: `False`.
+    channels_first : bool, optional
+        Whether the encoded payload is stored as (nchan, nsample, npol),
+        rather than (nsample, nchan, npol).  Default: `True`.
+    """
+    _decoders = {
+        8: decode_8bit}
+    _encoders = {
+        8: encode_8bit}
+    _dtype_word = np.dtype('int8')
+
+    _sample_shape_maker = namedtuple('SampleShape', 'npol, nchan')
+
+    def __init__(self, words, header=None, sample_shape=(), bps=8,
+                 complex_data=False, channels_first=True):
+        if header is not None:
+            bps = header.bps
+            sample_shape = header.sample_shape
+            complex_data = header.complex_data
+            channels_first = header.channels_first
+        super(GUPPIPayload, self).__init__(words, sample_shape=sample_shape,
+                                           bps=bps, complex_data=complex_data)
+        self.channels_first = channels_first
+        # If channels first, _item_to_slices must act on per-channel words.  By
+        # resetting self._bpfs, we allow _item_to_slices to work unmodified.
+        self._true_bpfs = self._bpfs    # Save the true bpfs regardless.
+        if self.channels_first:
+            self._bpfs //= self.sample_shape.nchan
+
+    @classmethod
+    def fromfile(cls, fh, header=None, memmap=False, payload_nbytes=None,
```

Comment:
Both `fromfile` and `tofile` are likely (nearly) identical to `DADA`, so maybe add to the issue about possible refactoring.